README.md

```diff
@@ -2,10 +2,103 @@
+
+## Implementation
+
+The SMT has 4 node types that are used to construct the tree:
+
+- Inner Nodes
+- Extension Nodes
+- Leaf Nodes
+- Lazy Nodes
+
+### Inner Nodes
+
+Inner nodes represent a branch in the tree with two **non-nil** child nodes. The inner node has an internal `digest` which represents the hash of the child nodes concatenated hashes.
+
+### Extension Nodes
+
+Extension nodes represent a singly linked chain of inner nodes, with a single child. They are used to represent a common path in the tree and as such contain the path and bounds of the path they represent. The `digest` of an extension node is the hash of its path bounds, the path itself and the child nodes digest concatenated.
+
+### Leaf Nodes
+
+Leaf nodes store the full path which they represent and also the hash of the value they store. The `digest` of a leaf node is the hash of the leaf nodes path and value concatenated.
+
+The SMT stores only the hashes of the values in the tree, not the raw values themselves. In order to store the raw values in the underlying database the option `WithValueHasher(nil)` must be passed into the `NewSparseMerkleTree` constructor.
+
+### Lazy Nodes
+
+Lazy nodes represent uncached, persisted nodes, and as such only store the `digest` of the node. When a lazy node is accessed the node type will be determined and the full node type will be populated with any relevant fields such as its children and path.
+
+### Lazy Loading
+
+This library uses a cached, lazy-loaded tree structure to optimize performance. It optimises performance by not reading from/writing to the underlying database on each operation, deferring any underlying changes until the `Commit()` function is called.
+
+All nodes have a `persisted` field which signals whether they have been persisted to the underlying database or not. In practice this gives a large performance optimisation by working on cached data and not reading from/writing to the database on each operation. If a node is deleted from the tree it is marked as `orphaned` and will be deleted from the database when the `Commit()` function is called.
+
+Once the `Commit()` function is called the tree will delete any orphaned nodes from the database and write the key-value pairs of all the unpersisted leaf nodes' hashes and their values to the database.
+
+### Visualisation
+
+The following diagram shows a simplified version of how the tree can be visualised using the node types described above.
+
+```mermaid
```

Comment:
I know this a non-trivial request, but can we extend on this? I have a bunch of questions after seeing it:

1. What does a Lazy Node look like?
2. What does Path=?
3. Why do we only have Path in the left subtree but not the right?
3. Can you show what digest=?
5. Can you distinguish between what the digest and value is?

Might need > 1 diagram to avoid it from getting overly complex.

Reply:
Have added some more clarification here let me know if it helps - a couple more diagrams but not sure how to put the path one into a diagram currently just some good ol text